### ISSUE TITLE
did not found a value for key SHINYPROXY_WEBSOCKET_RECONNECTION_MODE

### ISSUE THREAD

**Issue by feddelegrand7** [openanalytics/shinyproxy#380]
Hi :), 

First, thanks for the great product. I've recently installed Shiny Proxy 2.6.0 as I wanted to implement the following features: 
```
stop-proxies-on-shutdown: false
recover-running-proxies: true
recover-running-proxies-from-different-config: true
default-webSocket-reconnection-mode: Auto
default-stop-proxy-on-logout: true
```
I've modified my configuration file accordingly but when I release a new version of an app, I'm getting the following error: 

**Error
Status code: 500

Message: did not found a value for key SHINYPROXY_WEBSOCKET_RECONNECTION_MODE**

Can someone help, please: 

Thanks in advance.

**Comment by LEDfan** [comment]
Hi

Thanks for opening this issue. I'm not able to reproduce the issue. I think this issue may happen when you upgrade ShinyProxy from an older version to 2.6.0 while having some older apps running.
Can you try stopping all your old apps, upgrade ShinyProxy and then launch new apps?

**Comment by feddelegrand7** [comment]
Hi @LEDfan, Thanks for your answer, 

I'll try to do it this afternoon and see if the error still occurs. I have also a question about an aspect that's unfortunately not documented yet, whenever I use the following properties: 

stop-proxies-on-shutdown: false
recover-running-proxies: true

I don't get the Status code: 500 error however my apps still crash due to an "authentification_required" error. Can we use the app recovery feature with the ldap authentication? 

Thanks again, 
Fodil

**Comment by LEDfan** [comment]
In order to stay authenticated, you should be using Redis in order to achieve session persistence.
You can find info on configuring redis here: https://shinyproxy.io/documentation/configuration/#session-persistence

I noticed this isn't mentioned on the app recovery page, I'll make sure to update that page to include this information. Sorry for making this not clear before.

**Comment by feddelegrand7** [comment]
Hi @LEDfan, 

Thank you again for your help. Our devops team has created a redis docker container so that we can have app persistence. I also stopped all docker containers before releasing the 2.6.1 version, unfortunately, I still get the `Error: did not found a value for key SHINYPROXY_WEBSOCKET_RECONNECTION_MODE` 

Best, 
Fodil

**Comment by LEDfan** [comment]
Hi @feddelegrand7 

Accidentally I encountered this issue in one of our deployments and was able to find the cause of it. This issue should only happen when the original image used for the app is changed, such as in this scenario:

1. start ShinyProxy with an app using image myapp
2. launch the app
3. create a new version of the `myapp` image, for example by re-building it. For the running container, Docker will no longer report its image as `myapp` but using its digest.
4. start shinyproxy
5. shinyproxy will be unable to find the spec for the image and when opening the app gives the reported error

The issue is that ShinyProxy tries to find the containerspec using the name of the docker image. If Docker no long reports this using the original image name, shinyproxy is unable to find the container spec.

This is already fixed in the current development version of ShinyProxy, since we made some improvements to app recovery. In the meantime you can use the docker image `openanalytics/shinyproxy-snapshot:2.6.1-SNAPSHOT-20220920.093506` or [this jar](https://cloud.openanalytics.eu/s/M2Yq25LRDFZJ3mJ) which contain a fix for the issue.

**Comment by feddelegrand7** [comment]
Hi @LEDfan, 

Thank you so much for the fix. I've tested it right now and it's working perfectly, the only issue that I've is that the app (after updating the config.yaml - a new release) get tremendously slower. For example, it freezes for a couple of seconds when I try to fetch some data from the app. Is it related to the size of the container? 

Thanks again.

**Comment by LEDfan** [comment]
Hi @feddelegrand7 

I never experienced such a situation where an app freezes. I guess it could be because the app tries to perform some blocking calculations in the background and R is unable to handle other things. It should not be related to the reconnect settings of shinyproxy (of course you can try disabling them and see whether it makes a difference, but it shouldn't be the case). Did you make any progress on this issue?

**Comment by feddelegrand7** [comment]
Hi @LEDfan, 
Thanks for your reply, unfortunately I didn't make any progress regarding this issue. This is what happens: 
1- The runs fluently, when I click on the "get data" button, the app immediately starts querying the data. 
2- I change the configuration.yaml file for the app (to release a new feature).
3- The app doesn't crash. 
4- Then, when I click the "get data" button again, nothing happens for ~10 seconds (the app freezes), and then the app starts fetching the data. 
5- I get the data, I click again on the "get data" button, I still have to wait ~10 seconds again. 

I didn't have time to investigate more thoroughly, but I'll do asap and keep you updated. 

Thanks again :) 
Fodil

**Comment by feddelegrand7** [comment]
We run Redis in a docker container (with a password). Can it be the source of the issue?

**Comment by innir** [comment]
@LEDfan We are suffering from this issue as well, any timeline when to expect a fixed version?

**Comment by LEDfan** [comment]
Hi, the original issue with `SHINYPROXY_WEBSOCKET_RECONNECTION_MODE` has been fixed in ShinyProxy 3.0.0. In fact we improved the app recovery, such that it no longer reads the original spec from the configuration file. Therefore, you can make any changes to the spec (or even remove it ) and the app will still work.

See: https://shinyproxy.io/documentation/app-recovery/
